field/TinyLang/Field/Parser.hs

```diff
@@ -0,0 +1,213 @@
+{-| A parser for the boolean language.  The concrete syntax is as follows:
+
+  val ::= T | F
+  fvar ::= [a-z][a-z0-9_]*
+  bvar ::=  '?'[a-z][a-z0-9_]*
+
+  expr ::= val
+           fvar
+           bvar
+           'not' expr
+           'neq0' expr
+           'neg' expr
+           'inv' expr
+           expr 'and' expr
+           expr 'or'  expr
+           expr 'xor' expr
+           expr == expr
+           expr + expr
+           expr - expr
+           expr * expr
+           expr / expr
+           'if' expr 'then' expr 'else' expr
+           (expr)
+
+  Things like 'and' denote keywords.
+
+  Precedence: 'not' > 'xor' > 'and' > 'or'  (but use parentheses anyway).
+  if-then-else has to be parenthesised unless it's at the very top.
+
+  FIXME: what about numeric operators?
+
+  The code is based on the tutorial at
+  https://markkarpov.com/megaparsec/parsing-simple-imperative-language.html
+
+  See also https://markkarpov.com/megaparsec/megaparsec.html
+-}
+
+module TinyLang.Field.Parser
+    ( parseExpr,
+      ExprWrapper(..)
+    ) where
+
+import           TinyLang.Field.Core
+--import           TinyLang.Field.Printer
+import           TinyLang.Prelude               hiding (many, try)
+import           TinyLang.Var
+
+import           Control.Applicative (pure)
+import           Control.Monad.Combinators.Expr as E
+import qualified Data.Map                       as M
+import           Text.Megaparsec
+import           Text.Megaparsec.Char
+import           TinyLang.Field.ParsableField    
+import           TinyLang.Field.ParserUtils
+
+-- A uniform type to contain exprs of type bool/numeric.  Maybe this
+-- could go in Core.hs;  do we really need it though?
+data ExprWrapper f =
+    BoolExpr (Expr f Bool)
+  | FieldExpr (Expr f (AField f))
```

Comment:
We have a similar one in `TinyLang.Field.Evaluator`:

```haskell
data SomeUniVal f = forall a. SomeUniVal (Uni f a) a
```

This one would be

```haskell
data SomeUniExpr f = forall a. SomeUniExpr (Uni f a) (Expr f a)
```

but yours is simpler, so feel free to keep it and move to `Core` under the `SomeExpr` name, if desired.

Reply:
Replaced `ExprWrapper` by `SomeUniExpr` in 966f7d3eca5908848e583b0a8cc15e7e1e7bb434.